utils/zerovec/src/map/mod.rs

```diff
@@ -0,0 +1,272 @@
+    /// map.insert(1, "one".to_owned());
+    /// map.insert(2, "two".to_owned());
+    /// assert_eq!(map.contains_key(&1), true);
+    /// assert_eq!(map.contains_key(&3), false);
+    /// ```
+    pub fn contains_key(&self, key: &K::NeedleType) -> bool {
+        self.keys.binary_search(key).is_ok()
+    }
+
+    /// Insert `value` with `key`, returning the existing value if it exists.
+    ///
+    /// ```rust
+    /// use zerovec::ZeroMap;
+    ///
+    /// let mut map = ZeroMap::new();
+    /// map.insert(1, "one".to_owned());
+    /// map.insert(2, "two".to_owned());
+    /// assert_eq!(map.get(&1), Some("one"));
+    /// assert_eq!(map.get(&3), None);
+    /// ```
+    pub fn insert(&mut self, key: K, value: V) -> Option<V> {
+        let key_needle = key.as_needle();
+        match self.keys.binary_search(key_needle) {
+            Ok(index) => Some(self.values.replace(index, value)),
+            Err(index) => {
+                self.keys.insert(index, key);
+                self.values.insert(index, value);
+                None
+            }
+        }
+    }
+
+    /// Remove the value at `key`, returning it if it exists.
+    ///
+    /// ```rust
+    /// use zerovec::ZeroMap;
+    ///
+    /// let mut map = ZeroMap::new();
+    /// map.insert(1, "one".to_owned());
+    /// map.insert(2, "two".to_owned());
+    /// assert_eq!(map.remove(&1), Some("one".to_owned()));
+    /// assert_eq!(map.get(&1), None);
+    /// ```
+    pub fn remove(&mut self, key: &K::NeedleType) -> Option<V> {
+        let idx = self.keys.binary_search(key).ok()?;
+        self.keys.remove(idx);
+        Some(self.values.remove(idx))
+    }
+
+    /// Appends `value` with `key` to the end of the underlying vector, returning
+    /// `key` and `value` _if it failed_. Useful for extending with an existing
+    /// sorted list.
+    /// ```rust
+    /// use zerovec::ZeroMap;
+    ///
+    /// let mut map = ZeroMap::new();
+    /// assert!(map.try_append(1, "uno".to_owned()).is_none());
+    /// assert!(map.try_append(3, "tres".to_owned()).is_none());
+    ///
+    /// let unsuccessful = map.try_append(3, "tres-updated".to_owned());
+    /// assert!(unsuccessful.is_some(), "append duplicate of last key");
+    ///
+    /// let unsuccessful = map.try_append(2, "dos".to_owned());
+    /// assert!(unsuccessful.is_some(), "append out of order");
+    ///
+    /// assert_eq!(map.get(&1), Some("uno"));
+    ///
+    /// // contains the original value for the key: 3
+    /// assert_eq!(map.get(&3), Some("tres"));
+    ///
+    /// // not appended since it wasn't in order
+    /// assert_eq!(map.get(&2), None);
+    /// ```
+    #[must_use]
+    pub fn try_append(&mut self, key: K, value: V) -> Option<(K, V)> {
+        if self.keys.len() != 0 {
+            if let Some(last) = self.keys.get(self.keys.len() - 1) {
+                if key.cmp_get(last) != Ordering::Greater {
+                    return Some((key, value));
+                }
+            }
+        }
```

Comment:
Suggestion (optional):

```suggestion
        if let Some(last) = self.keys.last() {
            if key.cmp_get(last) != Ordering::Greater {
                return Some((key, value));
            }
        }
```

which would involve adding `.last()` to the trait.  But, this enables the container to implement it possibly more efficiently.

Reply:
It's probably not going to be more efficient, this is almost certainly getting optimized. The zero-check exists because of underflow but the two checks would get elided